# Prueba Técnica - Analista Jr de Inversiones
## Insights Wealth Management - 14 de Marzo del 2026


## Ejercicio 1 — Manejo de Datos

El objetivo es obtener los balances por rango de fechas y tipo de portafolio,
de los usuarios cuyo asesor es `insightswm@gmail.com`.

**Secuencia de funciones:**

1. `join(Asesores, Usuarios, id_asesor, Asesores_usuarios)`
2. `filter(Asesores_usuarios, email = "insightswm@gmail.com")`
3. `drop(Asesores_usuarios, email & num_clientes & perfil_riesgo)`
4. `join(Asesores_usuarios, Portafolios, id_usuario, AU_portafolios)`
5. `filter(AU_portafolios, tipo_portafolio = <tipo_deseado>)`
6. `join(AU_portafolios, Balances, id_portafolio, AU_balances)`
7. `filter(AU_balances, fecha >= "fecha_inicio" & fecha <= "fecha_fin")`
8. `drop(AU_balances, id_asesor & id_usuario & id_portafolio)`

## Ejercicio 2 — Portafolios

### 2.1 Fórmulas teóricas

Sea **R** la matriz de retornos simulados [30,000 x 55] y **P** un vector de pesos [1 x 55]:

- **Retorno esperado:** `E[Rp] = P · mean(R)` → promedio ponderado de los retornos medios de cada activo
- **Volatilidad esperada:** `σ(Rp) = std(R · Pᵀ)` → desviación estándar de los retornos simulados del portafolio

### 2.2 Cálculo numérico

In [ ]:
import pandas as pd
import numpy as np

R  = pd.read_excel("prueba.xlsx", sheet_name="Matriz de Simulacion", header=0).values
P1 = pd.read_excel("prueba.xlsx", sheet_name="P1", header=0).iloc[0].values
P2 = pd.read_excel("prueba.xlsx", sheet_name="P2", header=0).iloc[0].values

def portfolio_stats(P, R, name):
    retornos_portafolio = R @ P
    retorno_esperado    = np.mean(retornos_portafolio)
    volatilidad         = np.std(retornos_portafolio)
    print(f"=== {name} ===")
    print(f"Retorno esperado : {retorno_esperado:.4f} ({retorno_esperado*100:.2f}%)")
    print(f"Volatilidad      : {volatilidad:.4f} ({volatilidad*100:.2f}%)")
    print()

portfolio_stats(P1, R, "Portafolio P1")
portfolio_stats(P2, R, "Portafolio P2")

=== Portafolio P1 ===
Retorno esperado : 0.0912 (9.12%)
Volatilidad      : 0.2062 (20.62%)

=== Portafolio P2 ===
Retorno esperado : 0.0584 (5.84%)
Volatilidad      : 0.1011 (10.11%)



### 2.2 Resultados

| Portafolio | Retorno Esperado | Volatilidad |
|---|---|---|
| P1 | 9.12% | 20.62% |
| P2 | 5.84% | 10.11% |

P1 ofrece mayor retorno esperado (9.12%) pero asume mayor riesgo (20.62% de volatilidad).
P2 es más conservador con menor retorno (5.84%) y menor volatilidad (10.11%).
La relación retorno/riesgo (Sharpe simplificado) es similar en ambos portafolios,
lo que sugiere que ninguno domina claramente al otro sin considerar el perfil del inversionista.

### 2.3 Comparativa Claude vs Gemini

**Pregunta:** ¿Cómo afecta una subida de tasas de la Fed al retorno y volatilidad
esperados de un portafolio de renta fija y uno de renta variable?

| Criterio | Claude | Gemini |
|---|---|---|
| **Profundidad** | Técnico y directo. Explica duración, tasa de descuento y canales de transmisión sin adornos | Muy completo. Añade perspectiva temporal (corto vs largo plazo) y contexto de mercado actual ("Efecto 2026") |
| **Precisión** | Preciso en los mecanismos clave: relación inversa bonos/tasas, impacto en valuación de acciones | Preciso y agrega matices útiles: reinversión a largo plazo, tres canales de impacto en renta variable |
| **Tono** | Formal y técnico, orientado a un analista | Más didáctico y conversacional, usa analogías ("freno de mano") accesibles para el cliente |
| **Estructura** | Respuesta en prosa estructurada por tipo de activo | Mejor estructurado visualmente: bullets, tabla resumen, sección de contexto actual |

**Conclusión:** Escogería **Gemini** para comunicación con clientes por su claridad
y estructura visual. Sin embargo, preferiría **Claude** para análisis interno
o técnico donde la precisión y concisión son más valiosas que la didáctica.
Para esta prueba, Gemini demuestra mejor capacidad de síntesis ejecutiva.

## Ejercicio 3 — Automatización de Retiros

### Lógica de decisión

- **REJECT:** account_status ≠ active, kyc_status ≠ verified, amount ≤ 0, duplicado,
  o aml_risk_tier = high con destino no whitelisted
- **HOLD:** Sin razón de rechazo pero con riesgo de liquidez, destino reciente,
  o velocidad urgente con riesgo AML medio/alto
- **APPROVE:** No cumple ninguna condición de REJECT ni HOLD

### Parámetros
- BUFFER_USD = 50
- RECENT_DEST_DAYS = 7
- DUP_WINDOW_MIN = 15

In [ ]:
import pandas as pd
import numpy as np
from datetime import timedelta

# Parámetros
BUFFER_USD       = 50
RECENT_DEST_DAYS = 7
DUP_WINDOW_MIN   = 15

SEVERITY = {
    "KYC_NOT_VERIFIED":                   100,
    "ACCOUNT_NOT_ACTIVE":                  95,
    "UNWHITELISTED_HIGH_AML":              90,
    "INVALID_AMOUNT":                      85,
    "DUPLICATE_REQUEST":                   70,
    "INSUFFICIENT_SETTLED_AFTER_BUFFER":   65,
    "INSUFFICIENT_AVAILABLE_AFTER_BUFFER": 55,
    "DEST_CHANGED_RECENTLY":               45,
    "URGENT_RISK_TIER":                    35,
}

# Cargar datos
requests = pd.read_excel("withdrawals.xlsx", sheet_name="withdrawal_requests")
snapshot = pd.read_excel("withdrawals.xlsx", sheet_name="account_snapshot")
destReg  = pd.read_excel("withdrawals.xlsx", sheet_name="destination_registry")

# Parse fechas
requests["created_at"]     = pd.to_datetime(requests["created_at"], utc=True)
destReg["last_changed_at"] = pd.to_datetime(destReg["last_changed_at"], utc=True)
snapshot["as_of"]          = pd.to_datetime(snapshot["as_of"], utc=True)

# Merge
df = requests.merge(snapshot, on="account_id", how="left")
df = df.merge(destReg[["destination_id","is_whitelisted","last_changed_at"]], on="destination_id", how="left")

# Duplicados
df = df.sort_values("created_at").reset_index(drop=True)
dup_set, is_dup = set(), []
for _, row in df.iterrows():
    key = (row["account_id"], row["amount"], row["destination_id"])
    is_duplicate = any(
        k[0]==key[0] and k[1]==key[1] and k[2]==key[2] and
        (row["created_at"] - k[3]) <= timedelta(minutes=DUP_WINDOW_MIN)
        for k in dup_set
    )
    is_dup.append(is_duplicate)
    dup_set.add((key[0], key[1], key[2], row["created_at"]))
df["is_duplicate"] = is_dup

# Lógica de decisión
as_of_map = snapshot.set_index("account_id")["as_of"].to_dict()

def decide(row):
    reject, hold = [], []
    if row.get("account_status") != "active":        reject.append("ACCOUNT_NOT_ACTIVE")
    if row.get("kyc_status") != "verified":          reject.append("KYC_NOT_VERIFIED")
    if pd.isna(row["amount"]) or row["amount"] <= 0: reject.append("INVALID_AMOUNT")
    if row["is_duplicate"]:                          reject.append("DUPLICATE_REQUEST")
    if row.get("aml_risk_tier") == "high" and row.get("is_whitelisted") == False:
        reject.append("UNWHITELISTED_HIGH_AML")
    if reject:
        top = max(reject, key=lambda r: SEVERITY[r])
        return "REJECT", top, SEVERITY[top]

    avail, settled, amount = row.get("available_cash", 9e9), row.get("settled_cash", 9e9), row["amount"]
    if (avail - amount)   < BUFFER_USD: hold.append("INSUFFICIENT_AVAILABLE_AFTER_BUFFER")
    if (settled - amount) < BUFFER_USD: hold.append("INSUFFICIENT_SETTLED_AFTER_BUFFER")

    snap_as_of   = as_of_map.get(row["account_id"])
    last_changed = row.get("last_changed_at")
    if pd.notna(last_changed) and pd.notna(snap_as_of):
        if (snap_as_of - last_changed).days <= RECENT_DEST_DAYS:
            hold.append("DEST_CHANGED_RECENTLY")

    if row.get("requested_speed") == "urgent" and row.get("aml_risk_tier") in ("medium","high"):
        hold.append("URGENT_RISK_TIER")

    if hold:
        top = max(hold, key=lambda r: SEVERITY[r])
        return "HOLD", top, SEVERITY[top]

    return "APPROVE", "OK", 0

results = [{"request_id": r["request_id"], "account_id": r["account_id"],
            "amount": r["amount"], **dict(zip(["decision","reason_code","severity"], decide(r)))}
           for _, r in df.iterrows()]

decisions_df = pd.DataFrame(results)
hold_df = decisions_df[decisions_df["decision"]=="HOLD"].sort_values("severity", ascending=False)

print(decisions_df["decision"].value_counts().to_string())
print("\n=== HOLD QUEUE (top 10) ===")
print(hold_df[["request_id","amount","reason_code","severity"]].head(10).to_string(index=False))

decision
APPROVE    55
HOLD       38
REJECT     35

=== HOLD QUEUE (top 10) ===
request_id   amount                       reason_code  severity
   W100029 27278.01 INSUFFICIENT_SETTLED_AFTER_BUFFER        65
   W100109  7110.68 INSUFFICIENT_SETTLED_AFTER_BUFFER        65
   W100054  6719.15 INSUFFICIENT_SETTLED_AFTER_BUFFER        65
   W100045 27013.79 INSUFFICIENT_SETTLED_AFTER_BUFFER        65
   W100022 11612.55 INSUFFICIENT_SETTLED_AFTER_BUFFER        65
   W100052  9835.49 INSUFFICIENT_SETTLED_AFTER_BUFFER        65
   W100005  6859.80 INSUFFICIENT_SETTLED_AFTER_BUFFER        65
   W100072  8669.22 INSUFFICIENT_SETTLED_AFTER_BUFFER        65
   W100058 10730.44 INSUFFICIENT_SETTLED_AFTER_BUFFER        65
   W100012 20942.37 INSUFFICIENT_SETTLED_AFTER_BUFFER        65


In [ ]:
# Exportar resultados
decisions_df.to_excel("withdrawal_decisions.xlsx", sheet_name="decisions", index=False)

with pd.ExcelWriter("withdrawal_decisions.xlsx", engine="openpyxl") as writer:
    decisions_df.to_excel(writer, sheet_name="decisions", index=False)
    hold_df.to_excel(writer, sheet_name="review_queue", index=False)

print("✅ withdrawal_decisions.xlsx exportado")
print(f"\nResumen: APPROVE={( decisions_df['decision']=='APPROVE').sum()} | HOLD={(decisions_df['decision']=='HOLD').sum()} | REJECT={(decisions_df['decision']=='REJECT').sum()}")

✅ withdrawal_decisions.xlsx exportado

Resumen: APPROVE=55 | HOLD=38 | REJECT=35


### Bono — Propuesta de Agentización

**Idea:** Convertir este motor de decisiones en un agente autónomo event-driven.

**Flujo propuesto:**
1. **Trigger:** Cada solicitud de retiro llega vía API REST al agente
2. **Enriquecimiento:** El agente consulta en tiempo real account_snapshot,
   destination_registry y pending_activity desde la base de datos
3. **Decisión:** Corre la lógica APPROVE/HOLD/REJECT automáticamente
4. **Acciones:**
   - APPROVE → ejecuta el retiro vía API del broker
   - REJECT → notifica al cliente con reason_code
   - HOLD → crea ticket en sistema de revisión manual (ej. Jira/Slack)
5. **Monitoreo:** Dashboard en tiempo real con métricas de decisiones

**Herramientas sugeridas:**
- **Orquestación:** Apache Kafka o AWS SQS para procesar solicitudes en cola
- **Agente:** LangChain + Claude API para casos ambiguos que requieran criterio
- **Base de datos:** PostgreSQL con índices en account_id y destination_id
- **Notificaciones:** Twilio (SMS/WhatsApp) o SendGrid (email) al cliente
- **Monitoreo:** Grafana + Prometheus para alertas en tiempo real

**Ventaja:** Escala a miles de solicitudes por minuto sin intervención humana,
reduciendo el tiempo de procesamiento de horas a segundos.

## Ejercicio 4 — Solución con Agente ACH

### 4.1.1 ¿Cómo funciona ACH en EE.UU.?

El sistema ACH (Automated Clearing House) es la red de pagos electrónicos de EE.UU.
El flujo completo es:

**Originator → ODFI → ACH Network → RDFI → Receiver**

1. **Originator:** El cliente o empresa que inicia la transacción (ej. Insights)
2. **ODFI (Originating Depository Financial Institution):** El banco de Insights
   que envía el archivo ACH a la red
3. **ACH Network:** Operada por NACHA. Procesa y enruta los archivos en lotes
   (Fedwire o EPN)
4. **RDFI (Receiving Depository Financial Institution):** El banco del cliente
   que recibe los fondos
5. **Receiver:** El cliente cuya cuenta es debitada o acreditada

**Tiempos de liquidación:**
| Tipo | Tiempo | Disponibilidad |
|---|---|---|
| ACH Estándar | 1-3 días hábiles | Siguiente día hábil típicamente |
| Same-Day ACH | Mismo día hábil | Mismo día si se envía antes de las 2:45pm ET |

**¿Cuándo recomendar Same-Day ACH a un cliente de Insights?**
Cuando el cliente necesita fondear su cuenta urgentemente para aprovechar
una oportunidad de inversión el mismo día, o cuando el monto es menor a
$1,000,000 (límite de Same-Day ACH).

### 4.1.2 Requisitos para fondear vía ACH

**Datos bancarios necesarios:**
- Nombre del titular de la cuenta
- Routing number (ABA) del banco
- Número de cuenta bancaria
- Tipo de cuenta (checking o savings)

**Proceso de verificación:**
- Micro-depósitos: Insights envía 2 depósitos pequeños (<$1) al banco del cliente,
  el cliente confirma los montos (2-3 días)
- Verificación instantánea: vía Plaid o similar, el cliente autoriza acceso
  de solo lectura a su banco (inmediato)

**Límites típicos:**
- ACH estándar: $25,000 - $100,000 por transacción según el broker
- Same-Day ACH: hasta $1,000,000 por transacción

**Tiempos de disponibilidad de fondos:**
- ACH estándar: 1-3 días hábiles
- Same-Day ACH: mismo día hábil

**3 razones comunes de rechazo y mensaje al cliente:**

| Código NACHA | Razón | Mensaje a cliente |
|---|---|---|
| R01 | Fondos insuficientes | "Tu banco rechazó el depósito por fondos insuficientes. Por favor verifica tu saldo e intenta nuevamente." |
| R02 | Cuenta cerrada | "El número de cuenta bancaria registrado está cerrado. Por favor actualiza tu información bancaria." |
| R03 | Sin cuenta / cuenta inválida | "No encontramos una cuenta activa con los datos proporcionados. Verifica tu routing y número de cuenta." |

### 4.1.3 Routing Numbers (ABA)

Un **ABA Routing Number** es un código de 9 dígitos que identifica a una
institución financiera en EE.UU. Su estructura es:
- Dígitos 1-4: Código de la Reserva Federal
- Dígitos 5-8: Identificador del banco
- Dígito 9: Dígito verificador

**¿Por qué un banco puede tener routing distintos por estado?**
Los grandes bancos nacionales adquirieron bancos regionales a lo largo del tiempo.
Cada banco adquirido mantuvo su routing number original, por eso Bank of America
tiene routing distintos en California, Texas o Florida.

**Tabla de routing numbers — Bancos comunes para latinos en EE.UU.:**

| Banco | Florida | Texas | California | Nueva York | Illinois |
|---|---|---|---|---|---|
| Bank of America | 063100277 | 111000025 | 121000358 | 021000322 | 081904808 |
| Wells Fargo | 063107513 | 111900659 | 121042882 | 026012881 | 071101307 |
| Chase | 267084131 | 111000614 | 322271627 | 021000021 | 071000013 |
| Citibank | 266086554 | 113193532 | 322271724 | 021000089 | 271070801 |
| TD Bank | 067014822 | N/A | N/A | 026013673 | N/A |
| BBVA/PNC | 062001186 | 113010547 | N/A | N/A | 071921891 |
| Santander | N/A | N/A | N/A | 021407912 | N/A |
| Regions Bank | 062000019 | 065403626 | N/A | N/A | N/A |
| SunTrust/Truist | 061000104 | N/A | N/A | N/A | N/A |
| Popular/Banco Popular | 021571920 | N/A | N/A | 026015170 | N/A |

*Fuentes: sitios oficiales de cada banco y NACHA routing directory*

**¿Cómo el agente decide el routing cuando el cliente dice "Bank of America en Texas"?**
El agente usa una tabla de lookup interna (banco + estado → routing).
Si el cliente dice "Bank of America en Texas", el agente busca en la tabla
y responde con routing 111000025 sin que el cliente tenga que saberlo.
Si hay ambigüedad (banco no listado), el agente le pide al cliente que
verifique en el sitio oficial de su banco o en un cheque personal.

### 4.1.4 Comparativa ACH vs Wire vs Debit Card

| Criterio | ACH | Wire | Debit Card |
|---|---|---|---|
| **Velocidad** | 1-3 días (Same-Day disponible) | Mismo día (horas) | Inmediato |
| **Costo** | $0 - $3 | $15 - $50 | 1-3% del monto |
| **Límites típicos** | Hasta $1M | Sin límite práctico | $2,000 - $10,000 |
| **Reversibilidad** | Reversible hasta 60 días (R-codes) | Irrevocable | Reversible vía disputa |
| **Experiencia usuario** | Simple, solo datos bancarios | Requiere visitar banco | Muy simple |

**Conclusión:** ACH es el método preferido para Insights porque es de bajo costo,
familiar para el usuario y permite montos altos. Wire se recomienda para
transferencias urgentes de montos muy grandes. Debit Card para fondeos
pequeños e inmediatos de clientes nuevos.

### 4.2 Diseño del Agente

**Propósito:** Automatizar completamente la atención a clientes que desean
fondear su cuenta vía ACH, desde la recopilación de datos hasta la confirmación.

**Flujo conversacional:**
```
[INICIO]
    ↓
[1. Saludo e intención] → ¿Quieres fondear tu cuenta vía ACH?
    ↓
[2. Recopilación] → ¿Cuál es tu banco y en qué estado está registrado?
    ↓
[3. Lookup routing] → Inferir ABA routing number (banco + estado)
    ↓
[4. Datos adicionales] → Número de cuenta, tipo (checking/savings), monto
    ↓
[5. Instrucciones] → Guiar paso a paso el proceso de fondeo
    ↓
[6. Confirmación] → ¿Deseas confirmar la transferencia?
    ↓
[7. Resultado] → ÉXITO / FALLO (R01, R03) → Escalamiento
```

**Estados del agente:**
| Estado | Descripción |
|---|---|
| `recopilacion` | Pregunta banco y estado |
| `lookup_routing` | Infiere routing number |
| `instrucciones` | Guía el proceso paso a paso |
| `confirmacion` | Solicita confirmación final |
| `fallo` | Maneja errores R01/R03 y escala |

**Regla obligatoria:** El agente SIEMPRE pregunta banco y estado antes
de dar cualquier información de routing.

### 4.3 System Prompt
```text
Eres el asistente virtual de Insights Wealth Management, especializado en
ayudar a clientes a fondear su cuenta de inversión vía ACH.

REGLAS ESTRICTAS:
1. SIEMPRE pregunta el banco y el estado ANTES de dar cualquier routing number
2. Nunca inventes routing numbers — usa solo la tabla de lookup autorizada
3. Si el banco no está en la tabla, pide al cliente que lo verifique en su banco
4. Mantén un tono profesional, cálido y claro en todo momento
5. Si detectas un error (R01/R03), explica la causa y ofrece solución

FLUJO OBLIGATORIO:
1. Saluda e identifica la intención del cliente
2. Pregunta: banco + estado (OBLIGATORIO antes de cualquier dato)
3. Infiere el routing number con la tabla interna
4. Solicita: número de cuenta, tipo de cuenta, monto a fondear
5. Proporciona instrucciones paso a paso
6. Confirma los datos con el cliente
7. Simula el resultado (éxito o fallo)

MANEJO DE FALLOS:
- R01 (fondos insuficientes): "Tu banco rechazó la transferencia por fondos
  insuficientes. Verifica tu saldo e intenta de nuevo."
- R03 (cuenta inválida): "Los datos bancarios no corresponden a una cuenta activa.
  Verifica tu routing y número de cuenta."
- Si hay 2 fallos consecutivos: escalar a asesor humano

DISCLAIMER:
Insights Wealth Management no almacena tus datos bancarios completos.
Esta conversación es solo orientativa — la transferencia real se procesa
de forma segura a través de nuestro sistema encriptado.
```

### 4.4 Implementación del Agente ACH

El agente fue construido en Python usando la API de Gemini (Google AI) como
motor de lenguaje. Simula una conversación completa con un cliente que desea
fondear su cuenta vía ACH, cubriendo los flujos de éxito y fallos R01/R03.

**Para ejecutar localmente:**
1. Instalar dependencia: `pip install google-genai`
2. Obtener API key gratuita en: [aistudio.google.com](https://aistudio.google.com)
3. Reemplazar `YOUR_API_KEY` en el código por tu key personal
4. Ejecutar la celda en Colab o localmente con: `python agente_ach.py`

In [ ]:
!pip install google-genai -q

from google import genai
from google.genai import types

API_KEY = "AIzaSyClXQZLdT_--Pg28MQHzoE3EvU1sCOin5s"
client = genai.Client(api_key=API_KEY)

ROUTING_TABLE = {
    ("bank of america", "florida"):    "063100277",
    ("bank of america", "texas"):      "111000025",
    ("bank of america", "california"): "121000358",
    ("bank of america", "new york"):   "021000322",
    ("wells fargo", "florida"):        "063107513",
    ("wells fargo", "texas"):          "111900659",
    ("wells fargo", "california"):     "121042882",
    ("wells fargo", "new york"):       "026012881",
    ("chase", "florida"):              "267084131",
    ("chase", "texas"):                "111000614",
    ("chase", "california"):           "322271627",
    ("chase", "new york"):             "021000021",
    ("citibank", "florida"):           "266086554",
    ("citibank", "texas"):             "113193532",
    ("citibank", "new york"):          "021000089",
    ("td bank", "florida"):            "067014822",
    ("td bank", "new york"):           "026013673",
    ("popular", "florida"):            "021571920",
    ("popular", "new york"):           "026015170",
}

SYSTEM_PROMPT = """Eres el asistente virtual de Insights Wealth Management,
especializado en ayudar a clientes a fondear su cuenta de inversión vía ACH.

REGLAS:
1. SIEMPRE pregunta banco y estado ANTES de dar cualquier routing number
2. Usa solo routing numbers proporcionados por el sistema, nunca los inventes
3. Tono profesional, cálido y claro en todo momento

FLUJO:
1. Saluda e identifica la intención del cliente
2. Pregunta banco + estado (obligatorio antes de cualquier dato)
3. Confirma el routing number que el sistema te proporciona
4. Solicita número de cuenta, tipo de cuenta y monto
5. Da instrucciones paso a paso
6. Confirma datos y simula resultado

FALLOS:
- R01: Fondos insuficientes — sugiere verificar saldo e intentar de nuevo
- R03: Cuenta inválida — sugiere verificar routing y número de cuenta
- Después de 2 fallos consecutivos: escalar a asesor humano

DISCLAIMER: Insights no almacena datos bancarios completos."""

def lookup_routing(mensaje):
    for (banco, estado), routing in ROUTING_TABLE.items():
        if banco in mensaje.lower() and estado in mensaje.lower():
            return f"\n[SISTEMA: Routing para {banco.title()} en {estado.title()}: {routing}]"
    return ""

def run_agent():
    historial = []
    print("=" * 60)
    print("   Insights Wealth Management — Agente ACH")
    print("=" * 60)
    print("Escribe 'salir' para terminar\n")

    while True:
        entrada = input("Cliente: ").strip()
        if entrada.lower() == "salir":
            print("\nAgente: ¡Hasta luego!")
            break

        mensaje_completo = entrada + lookup_routing(entrada)
        historial.append(types.Content(role="user", parts=[types.Part(text=mensaje_completo)]))

        try:
            respuesta = client.models.generate_content(
                model="gemini-2.0-flash",
                config=types.GenerateContentConfig(system_instruction=SYSTEM_PROMPT),
                contents=historial
            )
            texto = respuesta.text
            historial.append(types.Content(role="model", parts=[types.Part(text=texto)]))
            print(f"\nAgente: {texto}\n")
        except Exception as e:
            print(f"\n[ERROR]: {e}\n")
            break

run_agent()

### 4.5 Transcripción de conversaciones

> El agente fue probado localmente con la API de Gemini. Se presentan 3
> escenarios representativos del flujo conversacional.

---

#### Escenario 1: Fondeo exitoso

**Cliente:** Hola buenas, quiero meter plata a mi cuenta de inversiones.

**Agente:** ¡Hola! Bienvenido a Insights. Para ayudarte a fondear tu cuenta
vía ACH necesito saber primero: ¿en qué banco tienes la cuenta y en qué
estado está registrada?

**Cliente:** Bank of America, tengo la cuenta en Texas.

**Agente:** Perfecto, el routing number para Bank of America en Texas es
111000025. Ahora dime: ¿cuál es tu número de cuenta, es checking o savings,
y cuánto quieres transferir?

**Cliente:** Es checking, número 987654321, quiero pasar $10,000.

**Agente:** Listo, te confirmo los datos antes de proceder:
- Banco: Bank of America (Texas)
- Routing: 111000025
- Cuenta: 987654321 — Checking
- Monto: $10,000 USD

¿Todo correcto?

**Cliente:** Sí, está bien.

**Agente:** Tu solicitud fue enviada. Los fondos deberían reflejarse en tu
cuenta de Insights en 1 a 3 días hábiles. Te llega confirmación al correo.
¿Hay algo más en lo que te pueda ayudar?

---

#### Escenario 2: Rechazo por fondos insuficientes (R01)

**Cliente:** Hola, quiero fondear mi cuenta.

**Agente:** Hola, con gusto. ¿En qué banco tienes tu cuenta y en qué estado?

**Cliente:** Chase, California.

**Agente:** El routing para Chase en California es 322271627. ¿Número de
cuenta, tipo y monto?

**Cliente:** Checking, 123456789, mando $50,000.

**Agente:** Hubo un problema con tu transferencia. Tu banco la rechazó con
código R01, que indica fondos insuficientes en la cuenta de origen. Te
recomiendo revisar tu saldo disponible antes de volver a intentarlo.
¿Quieres intentar con un monto menor?

**Cliente:** Bueno intentemos con $5,000.

**Agente:** Lo siento, tu banco rechazó la transferencia nuevamente por
el mismo motivo. En estos casos lo mejor es que uno de nuestros asesores
te contacte directamente para revisar la situación. Te voy a escalar el
caso y alguien te escribe en los próximos 30 minutos.

---

#### Escenario 3: Rechazo por cuenta inválida (R03)

**Cliente:** Buenas, necesito fondear mi cuenta.

**Agente:** Hola, claro que sí. ¿En qué banco y estado tienes tu cuenta?

**Cliente:** Wells Fargo en Florida.

**Agente:** El routing para Wells Fargo en Florida es 063107513. ¿Número
de cuenta, tipo y monto?

**Cliente:** Savings, 000111222, quiero pasar $15,000.

**Agente:** Tuve un problema al procesar tu solicitud. Los datos que
ingresaste no corresponden a una cuenta activa en Wells Fargo, lo cual
genera un rechazo R03. Puede ser un dígito mal escrito o que la cuenta
esté cerrada. ¿Puedes verificar el número en un estado de cuenta o en
la app del banco?

**Cliente:** Ah sí, el número correcto es 000111333.

**Agente:** Lo intenté de nuevo pero el sistema sigue sin reconocer esa
cuenta. Para no arriesgar tu dinero voy a escalar esto con un asesor que
te puede ayudar a verificar los datos directamente. Te contactan en menos
de 30 minutos.

---

### Reflexión

### Reflexión

Construir este agente me dejó varias conclusiones prácticas. Lo que más
me llamó la atención fue lo crítico que es controlar exactamente qué
información recibe el modelo. Si no se inyecta el routing number desde
una tabla validada, el LLM lo inventa con total confianza — en un contexto
financiero eso sería un error grave y potencialmente costoso para el cliente.

El flujo conversacional funcionó bien para los casos de éxito, pero los
fallos son donde más se nota la limitación del enfoque: el agente no tiene
acceso real al estado de la cuenta bancaria del cliente, así que la detección
de R01 o R03 depende de una integración externa con el banco o el broker.
Sin esa integración, el agente solo puede reaccionar a lo que el sistema
le informa, no detectarlo por sí solo.

Con más tiempo le agregaría tres cosas: memoria entre sesiones para no
preguntar banco y estado en cada conversación, integración con Plaid para
verificación instantánea de cuentas antes de enviar la solicitud, y manejo
de más códigos NACHA — actualmente solo cubre R01 y R03 pero existen más
de 80 códigos posibles. También mejoraría el escalamiento: en vez de solo
avisar al cliente, abriría un ticket automático en el CRM con el historial
completo de la conversación para que el asesor tenga contexto inmediato.